numpy/lib/arraypad.py

```diff
@@ -885,105 +885,81 @@ def _pad_wrap(arr, pad_amt, axis=-1):
-        ((before, after),) yields same before and after constants for each
-        axis.
-        (constant,) or int is a shortcut for before = after = constant for all
-        axes.
+    x : {None, scalar, array-like}
+        The object to broadcast to the shape (`ndim`, 2). None is only allowed
+        as a special case if `as_index` is True.
+    ndim : int
+        Number of pairs the broadcasted `x` will have.
+    as_index : bool, optional
+        If `x` is not None, try to round each element of `x` to a non-negative
+        integer.
+    assert_number : bool, optional
+        Raise a TypeError if the dtype of `x` is not a subdtype of `np.number`.
 
     Returns
     -------
-    _validate_lengths : tuple of tuples
-        int                               => ((int, int), (int, int), ...)
-        [[int1, int2], [int3, int4], ...] => ((int1, int2), (int3, int4), ...)
-        ((int1, int2), (int3, int4), ...) => no change
-        [[int1, int2], ]                  => ((int1, int2), (int1, int2), ...)
-        ((int1, int2), )                  => ((int1, int2), (int1, int2), ...)
-        [[int ,     ], ]                  => ((int, int), (int, int), ...)
-        ((int ,     ), )                  => ((int, int), (int, int), ...)
-
+    pairs : nested structure with shape (`ndim`, 2)
+        The broadcasted version of `x`.
+
+    Raises
+    ------
+    TypeError
+        If `as_index` is True and `x` is not None and can't be rounded to an
+        array of integer type.
+        Or if `assert_number` is True and the dtype of `x` is not a subdtype
+        of `np.number`.
+    ValueError
+        If `as_index` is True and `x` contains negative elements.
+        Or if `x` is not broadcastable to the shape (`ndim`, 2).
     """
-    normshp = _normalize_shape(narray, number_elements)
-    for i in normshp:
-        chk = [1 if x is None else x for x in i]
-        chk = [1 if x >= 0 else -1 for x in chk]
-        if (chk[0] < 0) or (chk[1] < 0):
-            fmt = "%s cannot contain negative values."
-            raise ValueError(fmt % (number_elements,))
-    return normshp
+    if as_index and x is None:
+        # Pass through None as a special case for indices
+        return ((None, None),) * ndim
+
+    x = np.asarray(x)
+
+    if x.ndim > 2:
+        # Fail explicitly when `x` already has more dimensions than
+        # desired shape (ndim, 2) accounts for
+        raise ValueError("keyword argument must not have more than 2 "
+                         "dimensions, got {}".format(x.ndim))
+    if assert_number and not np.issubdtype(x.dtype, np.number):
+        raise TypeError("keyword argument must be subdtype of np.number")
+
+    if as_index:
+        try:
+            x = x.round().astype(np.intp, copy=False)
+        except AttributeError:
+            raise TypeError("can't cast `x` to int")
+
+    if x.size == 1:
+        # Single value case
+        x = x.ravel()  # Reduce superfluous dimensions
```

Comment:
Why do we need this special case? `np.broadcast_to` handles it just fine.

Reply:
Knowing that `x.size == 1` (or `x.size == 2`) furthermore allows to test `as_index and x < 0` instead of the slower `as_index and x.min() < 0`.

Reply:
> Is there a reason why this may be a disadvantage?

I see no reason to be forgiving here - if the user passes a `(1, 1, 1, 1)`-shape array, we should tell them rather than letting them get away with it.